walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -17,59 +17,91 @@
-      return withoutOptionalAttendees;
-    } else {
-      return withOptionalAttendees;
-    }
-  }
-
-  private Collection<TimeRange> findAvailableTimes(
-      Collection<Event> events, MeetingRequest request, boolean withOptionalAttendees) {
     List<Event> eventsList = new ArrayList<Event>(events);
     Collections.sort(eventsList, Event.ORDER_BY_START);
     eventsList.add(null); // Run for loop additional time to account for EOD
+    long duration = request.getDuration();
 
-    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
-    Set<String> requestAttendees = new HashSet<String>(request.getAttendees());
-    if (withOptionalAttendees) {
-      requestAttendees.addAll(request.getOptionalAttendees());
-    }
+    List<TimeRange> mandatoryTimes = new ArrayList<TimeRange>();
+    List<TimeRange> optionalTimes = new ArrayList<TimeRange>();
+
+    Set<String> mandatoryAttendees = new HashSet<String>(request.getAttendees());
+    Set<String> optionalAndMandatoryAttendees = new HashSet<String>(request.getOptionalAttendees());
+    optionalAndMandatoryAttendees.addAll(mandatoryAttendees);
 
-    TimeRange prevEventTime = null;
-    TimeRange currEventTime = null;
+    HashMap<String, TimeRange> prevTime = new HashMap<String, TimeRange>();
+    prevTime.put(MANDATORY, null);
+    prevTime.put(OPTIONAL, null);
+    TimeRange currTime = null;
+    boolean addToMandatory = true;
+    boolean addToOptional = true;
     for (Event currEvent : eventsList) {
-      currEventTime = null;
+      currTime = null;
+      addToMandatory = true;
+      addToOptional = true;
       if (currEvent != null) {
-        currEventTime = currEvent.getWhen();
+        currTime = currEvent.getWhen();
         Set<String> currAttendees = currEvent.getAttendees();
-        if (Collections.disjoint(requestAttendees, currAttendees)) {
-          continue;
-        } else if (prevEventTime != null && prevEventTime.overlaps(currEventTime)) {
-          prevEventTime = combineOverlappingRanges(prevEventTime, currEventTime);
-          continue;
-        }
+        addToMandatory = markOverlappingAttendees(mandatoryAttendees, currAttendees)
+            && markOverlappingRanges(MANDATORY, prevTime, currTime);
+        addToOptional = markOverlappingAttendees(optionalAndMandatoryAttendees, currAttendees)
+            && markOverlappingRanges(OPTIONAL, prevTime, currTime);
       }
-      int start = (prevEventTime == null) ? TimeRange.START_OF_DAY : prevEventTime.end();
-      int end = (currEventTime == null) ? TimeRange.END_OF_DAY : currEventTime.start();
-      addTime(availableTimes, start, end, request.getDuration());
-      prevEventTime = currEventTime;
+      addToList(MANDATORY, mandatoryTimes, prevTime, currTime, duration, addToMandatory);
+      addToList(OPTIONAL, optionalTimes, prevTime, currTime, duration, addToOptional);
+    }
+
+    // Handle edge case where if there are no mandatory,
+    // then all optional are treated like mandatory
```

Comment:
This would be an interesting edge case (no need to fix it immediately, just food for thought):
Suppose we have a large event with 1000 optional attendees, and there's no time range that's available for everyone. In this case we want to find the time that works for the most people.